azure-active-directory-server/src/main/java/org/jetbrains/teamcity/aad/AADAuthenticationScheme.java

```diff
@@ -121,7 +122,12 @@ public HttpAuthenticationResult processAuthenticationRequest(@NotNull HttpServle
     final String email = token.getClaim(EMAIL_CLAIM);
 
     final ServerPrincipal principal = myPrincipalFactory.getServerPrincipal(name, oid, email, schemeProperties);
-
+   
+    if(principal == null)
```

Comment:
@jfatta current version of TeamCity core should throw exception in this case
throw new AuthenticationFailedException("Authenticated successfully, but failed to create new user since creating new users by login is not allowed", null, false);
so if login page does not display this message to the user it is a bug actual for all the auth schemes. I'd better fix it in core then make a workaround in particular scheme